### DISCUSSION TITLE
Problem with SEPLOS-V3 connection?

### DISCUSSION THREAD

**Discussion by Enterprise4400** [Louisvdw/dbus-serialbattery#984] [New BMS requests]
Hello team,

I have been using dbus-serialbattery with my two SEPLOS-V2 for 6 months.
Works wonderfully.
Now I got two more batteries.
These are SEPLOS-V3.
Dbus-serialbattery cannot find it.
I also replaced the RS-485 adapter from the SEPLOS-V2.
The RS-485 communicates with the SEPLOS TOOL V3.
what do I have to do so that the two SEPLOS-V3 are also displayed in the Victron Cerbo?

Thanks

**Comment by mdkeil** [comment]
I think v3 is not supported yet.

**Comment by mr-manuel** [comment]
Probably they have another protocol. You have to request it from your seller or from the manufacturer and adapt the driver.

**Comment by Enterprise4400** [comment]
Thank you for your answers.
Does anyone know whether the SEPLOS V3 protocol is still integrated?
I don't think I can change the Chinese's mind, persuade them to write the old protocol with the new information again.

**Comment by wollew** [comment]
AFAIK nobody's currently working on a V3 implementation. If you can convince Seplos to release protocol documentation that would be the first step. Depending on how different from V2 it is one could try to tackle an adaption, even without owning a V3 myself.

**Comment by marcelrv** [comment]
I also found it did not work..
I do see this: https://github.com/syssi/esphome-seplos-bms/blob/main/esp32-seplos-v3-example.yaml
at the bottom they refer to https://github.com/syssi/esphome-seplos-bms/blob/main/docs/XZH%20BMS%20Modbus-RTU%20Protocol.pdf

I may give that a try if it works, likewise if I have bit of time, I'll check/log  the communication between the new app and the BMS

**Reply by wollew** [reply]
So V3 is Modbus RTU instead of Modbus ASCII in the V2. But having some documentation it should be possible to implement some kind of support, at least if someone who owns a V3 is willing to do some testing and logging with their BMS.

**Reply by Enterprise4400** [reply]
Hello wollew,
I have two new SEPLOS V3.
Would be happy to help with the tests.

**Comment by marcelrv** [comment]
Yes, for sure

I logged yesterday the traffic and I see the regular Seplos software is making the following modbus RU request, which I've replayed in python.

For the majority I can see directly what parameter is linked, but esp for the 100+ coils it will take some effort to find the related exact parameter.

Anyway, here the seplos requests + responses for my device:
```
client.read_input_registers(0x1700, 0x33, slave=0):
[22618, 18477, 17772, 25955, 21605, 25448, 8259, 28462, 11340, 29796, 16973, 21297, 13907, 12848, 12353, 11603, 20528, 13634, 0, 0, 12595, 21328, 12341, 16946, 13105, 12594, 13872, 12342, 12544, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
585a482d456c65635465636820436f2e2c4c7464424d53313653323030412d5350303542003133535030354232333131323630303631000000000000000000000000
XZH-ElecTech Co.,LtdBMS16S200A-SP05B13SP05B2311260061

=========================================
client.read_input_registers(0x1301, 0x01, slave=0):
[16]
10
►

=========================================
client.read_input_registers(0x1000, 0x12, slave=0):
[5345, 65148, 29944, 30400, 36, 985, 1000, 1, 3340, 2836, 3343, 3339, 2840, 2832, 0, 180, 180, 1000]
14e1fe7c74f876c0243d93e81d0cb14d0fd0bb18b100b4b43e8


=========================================
client.read_input_registers(0x1100, 0x1a, slave=0):
[3342, 3343, 3341, 3340, 3343, 3342, 3341, 3339, 3339, 3342, 3339, 3340, 3339, 3340, 3340, 3343, 2839, 2833, 2832, 2840, 2731, 2731, 2731, 2731, 2843, 2831]
d0ed0fd0dd0cd0fd0ed0dd0bd0bd0ed0bd0cd0bd0cd0cd0fb17b11b10b18aabaabaabaabb1bb0f

=========================================
client.read_coils(0x1200, 0x90, slave=0):
0x0 False
0x1 False
0x2 False
0x3 False
0x4 False
0x5 False
0x6 False
0x7 False
0x8 False
0x9 False
0xa False
0xb False
0xc False
0xd False
0xe False
0xf False
0x10 False
0x11 False
0x12 False
0x13 False
0x14 False
0x15 False
0x16 False
0x17 False
0x18 False
0x19 False
0x1a False
0x1b False
0x1c False
0x1d False
0x1e False
0x1f False
0x20 False
0x21 False
0x22 False
0x23 False
0x24 False
0x25 False
0x26 False
0x27 False
0x28 False
0x29 False
0x2a False
0x2b False
0x2c False
0x2d False
0x2e False
0x2f False
0x30 False
0x31 False
0x32 False
0x33 False
0x34 False
0x35 False
0x36 False
0x37 False
0x38 False
0x39 False
0x3a False
0x3b False
0x3c False
0x3d False
0x3e False
0x3f False
0x40 True
0x41 False
0x42 False
0x43 False
0x44 False
0x45 False
0x46 False
0x47 False
0x48 False
0x49 False
0x4a False
0x4b False
0x4c False
0x4d False
0x4e False
0x4f False
0x50 False
0x51 False
0x52 False
0x53 False
0x54 False
0x55 False
0x56 False
0x57 False
0x58 False
0x59 False
0x5a False
0x5b False
0x5c False
0x5d False
0x5e False
0x5f False
0x60 False
0x61 False
0x62 False
0x63 False
0x64 False
0x65 False
0x66 False
0x67 False
0x68 False
0x69 False
0x6a False
0x6b False
0x6c False
0x6d False
0x6e False
0x6f False
0x70 False
0x71 False
0x72 False
0x73 False
0x74 False
0x75 False
0x76 False
0x77 False
0x78 True
0x79 True
0x7a False
0x7b False
0x7c False
0x7d False
0x7e False
0x7f False
0x80 False
0x81 False
0x82 False
0x83 False
0x84 False
0x85 False
0x86 False
0x87 False
0x88 False
0x89 False
0x8a False
0x8b False
0x8c False
0x8d False
0x8e False
0x8f False
```

**Reply by wollew** [reply]
Which modbus client lib are you using? Using slave=0 is sending to the broadcast address, which according to the standard does not send any reply data (if I understand it correctly). And at least minimalmodbus seems to comply to this, so we'll an address to send to (it should be visible in some debug output most likely).

**Reply by marcelrv** [reply]
ah, yes, I use pymodbus https://github.com/pymodbus-dev/pymodbus
I was mimicing the way the seplos app is requesting the data from the battery. (as I first did a wireshark trace)

I just now quickly tried to send with another slaveId (like 1,2,3,4 etc) but than I don't get any response from the BMS.

**Comment by marcelrv** [comment]
I made some more testing and a clear description of the addresses and their meaning.
See here:
https://github.com/marcelrv/seplosBMSv3/blob/main/README.md

**Reply by wollew** [reply]
This together with the [linked PDF ](https://github.com/syssi/esphome-seplos-bms/blob/main/docs/XZH%20BMS%20Modbus-RTU%20Protocol.pdf) should be sufficient to do an initial implementation. I'll see what I can do, but due to personal constraints this is not going to happen in the next few days.

**Comment by marcelrv** [comment]
![Screenshot_2024-03-09-13-34-30-376_com android chrome](https://github.com/Louisvdw/dbus-serialbattery/assets/5871213/e2f1a0bc-2744-40bb-a066-e537e1b2f849)
I made some progress yesterday by changing the broadcast slaveid in minimalmosbus.
At least i have communication.

**Reply by wollew** [reply]
So the screenshot looks like you've basically already implemented the SeplosV3 "driver" for serialbattery, or is my interpretation wrong?

**Comment by marcelrv** [comment]
nope you are right, make good progress.
Still need to fix the alarms and a few fields. I'll share a first version once that is done

![image](https://github.com/Louisvdw/dbus-serialbattery/assets/5871213/82ae9226-41ad-4aee-afef-bc7b9e9f6135)
![image](https://github.com/Louisvdw/dbus-serialbattery/assets/5871213/0cbeb979-b112-4091-83c1-3c4243a33ffc)
![image](https://github.com/Louisvdw/dbus-serialbattery/assets/5871213/242d69de-7037-4c70-990b-16cd3610166c)

**Reply by Enterprise4400** [reply]
Wow Marcelrv,
looks great!
I'm already looking forward to the first version.

**Comment by marcelrv** [comment]
For those interested or want to test as well.... alarms now also work. I put my work in progress here..
https://github.com/marcelrv/dbus-serialbattery/tree/seplosv3

TODO:
* Set the fet status according to the pulled data (IO submenu). currently hardcoded!
* Check what variables are driving the parameters submenu, than update those correctly
* Test the alarms (to make sure it properly works for all the alarms)
* modbus address is currently hardcoded to 0 (which requires the minimalmodbus _SLAVEADDRESS_BROADCAST to be modified to anything except 0). Test if this also works by setting the dip on the BMS to anything else than 0
* If above works, add the identification of batteries on other addresses than 0
* Cleanup and removal of unnessesary logging

![image](https://github.com/Louisvdw/dbus-serialbattery/assets/5871213/7212776f-958c-45c6-bb04-e55d8cb8ecfe)

**Reply by marcelrv** [reply]
@Enterprise4400 your input/testing would be interesting as you indicated you have multiple batteries.
If I understand well, the address 0 would be the response of the master with the sum of the values from all the other batteries.
But that would be good if that is indeed confirmed.

**Reply by Enterprise4400** [reply]
Hello
marcelrv,
Yes, the master is 0.
Yes, exactly, I have two V3 SEPLOS and two V2 SEPLOS.
Please give me information when I can test?

**Reply by Enterprise4400** [reply]
Wow, the sceershots look great.
Just like you're used to.

**Reply by marcelrv** [reply]
It is now available for testing as it is merged.

I.m curious how the 2 battery's will work

**Comment by tnt760** [comment]
Hello all, first post.
Last night I hooked up my seplos v3 and can report back that my initial impressions are very positive.
Thankyou for the development.

**Reply by tugsi** [reply]
@tnt760 Did you also connect the two Seplos v3 to each other (master/slave) (RS485) or each one individually to the GX?
